rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,218 @@
+There's an argument then that pruning (and for that matter reverse sync) shouldn't 
+lie on the consensus critical path but rather be an asynchronous background process 
+of the node.
+
+Further to this argument, nodes will not be able to prune past the evidence age 
+regardless of the retain height that the application requests. Finally, applications 
+may ask to fetch blocks back to a height that none of the nodes peers have. 
+The mechanism must ensure termination even if it doesn't reach the height requested 
+by the app.
+
+Taking all these points into consideration, ReverseSync and pruning should be
+operations that run asynchronously in the background. Application can be informed
+of the current base via `RequestEndBlock`:
+
+```proto 
+message RequestEndBlock {
+  Height uint64
+  Base uint64
+}
+```
+
+### Verification
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
+```
+
+The node can therefore trust the new header. This then can be followed with
+using the hashes in the trusted header to validate the other block and state
+info. This is recursively done till the backfill height.  
+
+### Design
+
+This section will provide a high level picture of the design with the specifics being
+addressed in a following ADR. The reverse mechanism requires the gossiping of two new
+messages to allow the passsing of state data:
+
+```proto
+message StateDataRequest {
+  uint64 height = 1;
+}
+```
+
+```proto
+message StateDataResponse {
+  uint64 height = 1;
+  ValidatorSet validator_set = 2;
+  ConsensusParams consensus_params = 3;
+  ABCIResponses abci_responses = 4;
+}
+```
+
+This would be sent across a new channel most likely within the blockchain reactor 
+(one can view ReverseSync as a similar process to FastSync and thus should most
+likey be in the same domain).
+
+## Alternative Solutions
+
+If we do not want to extend the blockchain's functionality, it is possible to
+use the embedded light client within state sync instead to verify backwards outside
+of the unbondind period and request a snapshot at that height instead. 
+
+Taking one step further back, some of the problems listed in the introduction
+could be resolved by other means. Full nodes (and even validators to a certain
+extent) could simply bypass validation if they didn't have the necessary
+history trusting purely in consensus. Node operators running nodes with truncated 
+history could start up a separate full node if they wanted complete history.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Greater flexibility over the data that a node has. This makes it easier to
+become a node with full history.
+- Ensures a minimum block history invariant.
```

Comment:
I don't understand how this has been ensured, isn't it still up to honesty/benevolence assumptions? (Like nothing on the chain / consensus enforces this right?)

Reply:
Sure, the amount of history a node has isn't enforced on chain, but this means that no honest/correct node will be capable of pruning past a certain threshold which may cause their node to panic.

Reply:
That makes sense, could this be changed to 
` - Ensures a minimum block history invariant given a single correct node` to reflect this?